ext/statsd/statsd.c

```diff
@@ -0,0 +1,376 @@
+#include <stdbool.h>
+#include <ruby.h>
+#include <ruby/encoding.h>
+#include <ruby/st.h>
+
+#define DATAGRAM_SIZE_MAX 4096
+#define SAMPLE_RATE_SIZE_MAX 16
+#define NORMALIZED_TAGS_CACHE_ENABLED 1
+#define NORMALIZED_TAGS_CACHE_MAX 512
+#define NORMALIZED_NAMES_CACHE_ENABLED 1
+#define NORMALIZED_NAMES_CACHE_MAX 512
+
+static ID idTr, idNormalizeTags, idDefaultTags, idPrefix;
+
+struct datagram_builder {
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  st_table *normalized_tags_cache;
+#endif
+#ifdef NORMALIZED_NAMES_CACHE_ENABLED
+  st_table *normalized_names_cache;
+#endif
+  VALUE str_normalize_chars;
+  VALUE str_normalize_replacement;
+  // cached default tags ivar to skip a lookup
+  VALUE default_tags;
+  bool empty_default_tags;
+  int prefix_len;
+  int len;
+  // last member to not glob up cache lines to access other struct members
+  char datagram[DATAGRAM_SIZE_MAX];
+};
+
+// GC callback to mark the wrapper struct. Conditionally symbol tables if caching is enabled (values only)
+// and the cached default tags as well.
+void
+datagram_builder_mark(void *ptr)
+{
+  const struct datagram_builder *builder = (struct datagram_builder *)ptr;
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  rb_mark_tbl(builder->normalized_tags_cache);
+#endif
+#ifdef NORMALIZED_NAMES_CACHE_ENABLED
+  rb_mark_tbl(builder->normalized_names_cache);
+#endif
+  rb_gc_mark(builder->str_normalize_chars);
+  rb_gc_mark(builder->str_normalize_replacement);
+  rb_gc_mark(builder->default_tags);
+}
+
+// GC callback to free the wrapper struct. Conditionally symbol tables if caching is enabled
+// and the struct itself.
+void
+datagram_builder_free(void *ptr)
+{
+  struct datagram_builder *builder = (struct datagram_builder *)ptr;
+  if (!builder) return;
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  st_free_table(builder->normalized_tags_cache);
```

Comment:
Maybe set these to `NULL` after freeing?

Reply:
Yes agree, use after free guard

Reply:
https://github.com/Shopify/statsd-instrument/pull/259/commits/666ff836dd4f05fdd0c92b8ab267e10f3fd91e0a